docs/design/structs-interfaces.md

```diff
@@ -0,0 +1,304 @@
+
+An abandoned idea involved storing `Config` and `State` on the resource struct. This approach seems strictly worse than including config and state values in CRUD function request and response objects (e.g. `ResourceCreateRequest`), since it would be unclear when and how the framework modified those values.
+
+#### Interfaces
+
+In fulfilling an interface by creating instances of their own provider, resource, and data source struct types, provider developers have the **advantage** of more flexibility in their choice of fields within those structs, and whether they are used to store their own state.
+
+Allowing provider developers to define their own resource types, and construct them with a factory pattern, means that the type assertion on any meta parameter need be done only once, in the factory itself. This is an **advantage**.
+
+A **disadvantage** of the interface approach is that provider-defined structs are now responsible for managing state that framework helpers may need to access. An example of this is `TerraformVersion`, which is included in the `ConfigureProvider` RPC and needs to be available for inclusion in User-Agent HTTP headers via framework helpers that will be called from resource CRUD functions. Assuming `TerraformVersion` is to be stored on `Provider`, it can be set either when the struct is created, by passing arguments to the factory:
+
+```go
+func NewProvider(data framework.RuntimeState) Provider {
+  return provider{
+    terraformVersion: data.TerraformVersion,
+  }
+}
+```
+
+or with explicit methods:
+
+```go
+type provider struct {
+  framework.Provider
+  terraformVersion string
+}
+
+func (p *provider) SetTerraformVersion(version string) {
+  p.terraformVersion = version
+}
+```
+
+However the provider-controlled struct is given this runtime state, helpers controlled by the framework need access to `terraformVersion`, so a getter is also required:
+
+```go
+func (p provider) GetTerraformVersion() string {
+  return p.terraformVersion
+}
+```
+
+This implementation burden on the provider code can be mitigated by providing a framework-controlled struct that providers are required to embed in their interface implementations. This can be enforced at compile time by requiring an unexported method to be implemented.
+
+Framework code:
+```go
+type RuntimeData struct {
+  terraformVersion string
+  ConfigureProviderData interface{}
+}
+
+func (r RuntimeData) getRuntimeData() RuntimeData {
+  return r
+}
+```
+
+Provider code:
+```go
+type provider {
+  framework.RuntimeData
+}
+```
+
+`ConfigureProviderData` here represents other runtime data supplied by the `ConfigureProvider` request. Please see the document on Runtime State for further discussion of what runtime data may be injected and the tradeoffs involved.

```

Comment:
(I don't think this needs to be changed, but I just want to be clear about what I was intending: `ConfigureProviderData` in my example is a 1:1 mapping to `meta` in SDKv2. I think this comment should only result in a change if it changes your understanding of this paragraph. I think the paragraph is fine and accurate as-is, however.)